proposals/core.md

```diff
@@ -147,38 +150,127 @@ WebDriver has a notion of a current top-level browsing context, and current brow
+
+#### Windows Handles
 
-In this new model, a Target is simply a thing that can host some browsing contexts, and/or script contexts. A basic page target would host a single top-level browsing context, some nested browsing contexts, and a number of script contexts; one for each browsing context in the tree. There could be additional script contexts as well if, say, the page uses web workers. A service worker target would have no browsing contexts; it would host script contexts only. This is where splitting the notion of browsing and script contexts comes in handy; It is now possible to execute script in the context of a service worker.
+Today, WebDriver uses strings called "window handles" to uniquely identify top-level browsing contexts.
 
-To maintain backwards compatibility with classical WebDriver, Targets would act just like Windows and continue to use window handles (strings) as IDs. Service workers, and other types of targets that don't host any browsing contexts would be invisible to the old protocol. Calling the classic "Get Window Handles" command would return only the page Targets and attempting to switch to a non-page Target using a classic command would return an error.
+To obtain an HTTP-compatible window handle given a BiDi Window, the client can either query the BiDi Window for it's `httpWindowHandle` property, or get the `httWindowHandle` from the `targetCreated` event params when the Window is first created.
 
-*Open Issue: Do we want to add these concepts to the HTTP protocol as well?*
+To obtain a BiDi Window ID given an HTTP window handle, the client can either maintain a mapping of previously-seen BiDi Window IDs to window handles. The `UserAgent` can also provide a `findWindowByHttpHandle` command to perform the conversion.
 
-### Identifying elements
+#### Frames & Elements
 
-Web element references can continue to work as they do today, with one caveat; since element references are valid only within a particular browsing context, any commands that operate on elements and any events involving elements will need to specify which browsing context the element(s) belongs to. For example, a command like getElementText would need both a "browsingContextId" and an "elementId" parameter.
+WebDriver/HTTP uses Web Element References to represent HTML Elements and to represent Frames in the Switch To Frame command. However, Web Element References are valid only within a particular browsing context. To obtain a BiDi Element or Frame ID given a Web Element Reference, the client would first need to know which browsing context the Web Element Reference belongs to. The `BrowsingContext` can provide `findElementByHttpHandle` and `findBrowsingContextByHttpHandle` commands to convert Web Element Referenced into a BiDi Elements or Frames respectively.

```

Comment:
In the case of frames, I agree we need a new concept of a unique identifier (which will be just like the identifier for a tlbc), but it is also possible to simplify the compatibility story here by allowing the concept of a relative frame reference like `{type: "frame", "parent": <window id>, "ref": <webdriver/http reference>}`